src/aeso_ast_infer_types.erl

```diff
@@ -1466,15 +1466,34 @@ check_reserved_entrypoints(Funs) ->
 -spec check_fundecl(env(), aeso_syntax:decl()) -> {{name(), typesig()}, aeso_syntax:decl()}.
 check_fundecl(Env, {fun_decl, Ann, Id = {id, _, Name}, Type = {fun_t, _, _, _, _}}) ->
     Type1 = {fun_t, _, Named, Args, Ret} = check_type(Env, Type),
-    {{Name, {type_sig, Ann, none, Named, Args, Ret}}, {fun_decl, Ann, Id, Type1}};
+    TypeSig = {type_sig, Ann, none, Named, Args, Ret},
+    delete_if_implementation(Env, Name, TypeSig),
+    {{Name, TypeSig}, {fun_decl, Ann, Id, Type1}};
 check_fundecl(Env, {fun_decl, Ann, Id = {id, _, Name}, Type}) ->
     type_error({fundecl_must_have_funtype, Ann, Id, Type}),
     {{Name, {type_sig, Ann, none, [], [], Type}}, check_type(Env, Type)}.
 
+%% Delete the function FunName with the signature FunSig from of functions
+%% to be implemented if it is included there, or return ok otherwise.
+-spec delete_if_implementation(env(), FunName, FunSig) -> true when
+      FunName :: string(),
+      FunSig :: typesig().
+delete_if_implementation(Env, Name, Sig) ->
```

Comment:
Also, why do you match type sig as well?

Reply:
Changed here 2a55823633625f40d7143d48157e0844a976769c